agent/acs/handler/acs_handler.go

```diff
@@ -189,70 +189,61 @@ func NewSession(
 // Start starts the session. It'll forever keep trying to connect to ACS unless
 // the context is cancelled.
 //
-// If the context is cancelled, Start() would return with the error code returned
-// by the context.
-// If the instance is deregistered, Start() would emit an event to the
-// deregister-instance event stream and sets the connection backoff time to 1 hour.
+// If the context is cancelled returns nil otherwise returns error from context
 func (acsSession *session) Start() error {
-	// connectToACS channel is used to indicate the intent to connect to ACS
-	// It's processed by the select loop to connect to ACS
-	connectToACS := make(chan struct{}, 1)
-	// This is required to trigger the first connection to ACS. Subsequent
-	// connections are triggered by the handleACSError() method
-	connectToACS <- struct{}{}
+	// Loop continuously until context is closed/cancelled
 	for {
-		select {
-		case <-connectToACS:
-			seelog.Debugf("Received connect to ACS message")
-			// Start a session with ACS
-			acsError := acsSession.startSessionOnce()
-			select {
-			case <-acsSession.ctx.Done():
-				// agent is shutting down, exiting cleanly
+		seelog.Debugf("Attempting connect to ACS")
+		// Start a session with ACS
+		acsError := acsSession.startSessionOnce()
+
+		// If the session is over check for shutdown first
+		if err := acsSession.ctx.Err(); err != nil {
```

Comment:
I think this is a subtle behavior change, since before we were only checking `ctx.Done()` to trigger just exiting cleanly with a nil error. I believe `ctx.Err()` can return `context.DeadlineExceeded` in addition to `context.Canceled`.

I'm not sure why exactly we would ever receive `context.DeadlineExceeded` from this context in particular, but is there any reason in particular to change the previous behavior?

Reply:
I was aligning to the previous comment for the function. However I don't think it matters. If we just want to unconditionally exit then I would recommend getting rid of the error return value altogether since it is never provided.

That changes the interface and another test which mocks the interface to test when it returns an error which in reality is never. As can be seen in the tests I added checking the return at all so previously there was no validation of any behavior over time. 

It is a little more change but probably the most correct to remove the return to make it clear that this doesn't intend to do anything but return when the context is Done()

Reply:
> If we just want to unconditionally exit then I would recommend getting rid of the error return value altogether since it is never provided.

I'm not sure if I'm following exactly. This is what we currently do isn't it? ie, in this code here: https://github.com/aws/amazon-ecs-agent/blob/master/agent/acs/handler/acs_handler.go#L209-L214

My understanding of this PR is that you're changing from unconditionally exiting whenever the context is done, to conditionally exiting cleanly only when the err returned from the context is context.Canceled.

Reply:
> // If the context is cancelled, Start() would return with the error code returned
> // by the context.
^ this was what exists and I was trying to actually make true. This seems to be confusing the PR though so I will just remove that comment completely and go back to the previous behavior (even though in actual usage there is no difference since the context in use doesn't have a timer associated so can only return cancelled).

This PR is about removing brittle/unnecessary use of channels in the maintaining of an ACS connection. This https://github.com/aws/amazon-ecs-agent/blob/8dc9fdeb7b876dad609b06001448d0d04e4825fe/agent/app/agent.go#L915 is the only usage of the return value. Cancelled is the standard exit mechanism so printing the error in that case is strange, so I returned nil in that case.

Since outside of a test case with a mock acsSession you can never hit this log line in agent.go I was arguing that the error return value should be removed completely since it isn't used and IF at some point it was used it would probably need thought into what Agent.go did with it such that adding it back into that file at that time would be appropriate.

Will post new rev without these changes in any case. Removing the return value can be a separate PR.